### PR TITLE
Add equality comparison for Attributes.

### LINKED ISSUE #336: Add comparator for Attributes
This would be simple pointer/id comparison when attributes are canoncialized.

### PR DESCRIPTION
**Pull request by bgogul**
This partially fixes #336.

### CODE REVIEW

**Review by markww — APPROVED** [review]
Just nits, looks good otherwise.

File: src/ir/attributes/attribute.h [hunk 1]
```diff
@@ -67,17 +69,28 @@ class Attribute {
...
  private:
   // Private constructor that allows us to construct an attribute from an
   // intrusive_ptr to any derived type of `AttributeBase`.
   template <class T,
             std::enable_if_t<std::is_convertible<T*, AttributeBase*>::value,
                              bool> = true>
-  Attribute(const intrusive_ptr<const T>& value) : value_(value) {}
+  Attribute(const intrusive_ptr<const T>& value) : value_(value) {
+    CHECK(value_ != nullptr);
+  }
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/attributes/attribute.h]: Done.

File: src/ir/attributes/attribute.h [hunk 2]
```diff
@@ -67,17 +69,28 @@ class Attribute {
...
 
   intrusive_ptr<const AttributeBase> value_;
 };
 
+// TODO(#336): This will simply become a pointer comparison when we canonicalize
+// the attributes.
+inline bool operator==(const Attribute& lhs, const Attribute& rhs) {
+  CHECK(lhs.value_ != nullptr && rhs.value_ != nullptr);
+  return lhs.value_->IsEqual(*rhs.value_);
+}
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/attributes/attribute.h]: done.

File: src/ir/attributes/attribute_test.cc [hunk 3]
```diff
@@ -53,6 +53,33 @@ CreateTestAttribute() {
   return std::make_pair(Attribute::Create<StringAttribute>("Hello World!"),
                         StringAttribute::Create("Hello World!"));
 }
+struct AttributeWithId {
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/attributes/attribute_test.cc:57]: done.

File: src/ir/attributes/attribute.h [hunk 4]
```diff
@@ -67,17 +69,28 @@ class Attribute {
...
+  }
 
   intrusive_ptr<const AttributeBase> value_;
 };
 
+// TODO(#336): This will simply become a pointer comparison when we canonicalize
+// the attributes.
+inline bool operator==(const Attribute& lhs, const Attribute& rhs) {
+  CHECK(lhs.value_ != nullptr && rhs.value_ != nullptr);
+  return lhs.value_->IsEqual(*rhs.value_);
```
[the remark this answers is not in the document]
Comment (bgogul) [reply src/ir/attributes/attribute.h:92]: Yes, we need dynamic dispatch.

**Pull request closed without merging**